src/Config/Processor/BuilderProcessor.php

```diff
@@ -47,19 +47,29 @@ final class BuilderProcessor implements ProcessorInterface
      */
     public static function process(array $configs): array
     {
+        $addedTypes = [];
+
         foreach ($configs as &$config) {
             if (isset($config['config']['builders']) && \is_array($config['config']['builders'])) {
-                $buildersFields = self::processFieldsBuilders($config['config']['builders']);
+                ['fields' => $buildersFields, 'types' => $buildersTypes] = self::processFieldsBuilders($config['config']['builders']);
+
                 $config['config']['fields'] = isset($config['config']['fields']) ? \array_merge($buildersFields, $config['config']['fields']) : $buildersFields;
+
+                $addedTypes = \array_merge($addedTypes, $buildersTypes);
+
                 unset($config['config']['builders']);
             }
 
             if (isset($config['config']['fields']) && \is_array($config['config']['fields'])) {
-                $config['config']['fields'] = self::processFieldBuilders($config['config']['fields']);
+                ['fields' => $buildersFields, 'types' => $buildersTypes] = self::processFieldBuilders($config['config']['fields']);
+
+                $config['config']['fields'] = $buildersFields;
+
+                $addedTypes = \array_merge($addedTypes, $buildersTypes);
             }
         }
 
-        return $configs;
+        return \array_merge($configs, $addedTypes);
```

Comment:
Prevent $addedTypes from overriding types comming from $config. Throw in case of duplicate. At this point we do not know which builder it was. We could pass in to the processField*Builders methods a list of type names comming from $config, so they can check this. So they prevent builders override each others extra types and also the ones from $config and give clear message about what is the case and which builder it was.